src/inmanta/server/agentmanager.py

```diff
@@ -238,62 +238,107 @@ async def all_agents_action(self, env: data.Environment, action: AgentAction) ->
         elif action is AgentAction.keep_paused_on_resume:
             await self._set_unpause_on_resume(env, should_be_unpaused_on_resume=False, endpoint=name)
         elif action is AgentAction.unpause_on_resume:
             await self._set_unpause_on_resume(env, should_be_unpaused_on_resume=True, endpoint=name)
         else:
             raise BadRequest(f"Unknown agent action: {action.name}")
 
-    async def _pause_agent(
+    async def _stop_agent(
         self, env: data.Environment, endpoint: Optional[str] = None, connection: Optional[asyncpg.connection.Connection] = None
     ) -> None:
         """
         Pause a logical agent by pausing an active agent instance if it exists, and removing the logical agent's primary.
         """
-
+        endpoints_with_new_primary = []
         async with self.session_lock:
             agents = await data.Agent.pause(env=env.id, endpoint=endpoint, paused=True, connection=connection)
-            endpoints_with_new_primary = []
-            for agent_name in agents:
-                key = (env.id, agent_name)
+            if opt.server_use_resource_scheduler.get():
+                key = (env.id, const.AGENT_SCHEDULER_ID)
                 live_session = self.tid_endpoint_to_session.get(key)
-                if live_session:
-                    # The agent has an active agent instance that has to be paused
-                    del self.tid_endpoint_to_session[key]
-                    await live_session.get_client().set_state(agent_name, enabled=False)
-                    endpoints_with_new_primary.append((agent_name, None))
-            await data.Agent.update_primary(
-                env.id, endpoints_with_new_primary, now=datetime.now().astimezone(), connection=connection
-            )
+                # This should never be the case but as a safety measure, we need to make sure that the scheduler is up and
+                # running
+                if not live_session:
+                    await self._autostarted_agent_manager._ensure_scheduler(env)
+                    live_session = self.tid_endpoint_to_session.get(key)
+                    assert live_session
+
+                if endpoint is not None:
+                    # We don't need to do this when the environment is resumed because the scheduler will need to have updated
+                    # information related to agents (being paused) and in order to have that, this transaction needs to
+                    # finish first
+                    await live_session.get_client().set_state(endpoint, enabled=False)
+                elif endpoint is None:
+                    # We need to update information in the DB
+                    endpoints_with_new_primary.append((const.AGENT_SCHEDULER_ID, None))
+                    await data.Agent.update_primary(
+                        env.id, endpoints_with_new_primary, now=datetime.now().astimezone(), connection=connection
+                    )
```

Comment:
@wouterdb  I'm not really convinced by my implementation (the same can be said for the `start_agent` method): because if the scheduler is used, we cannot remove the live session here (as we would do if the scheduler wasn't used) because we still need to notify the scheduler that something has changed in the DB. But to do that, we need the transaction to go through. Any suggestion?